src/forest.rs

```diff
@@ -304,299 +304,537 @@ impl<'i, G: GrammarReflector, I: Input> ParseForest<'i, G, I> {
     }
 }
 
-// FIXME(rust-lang/rust#54175) work around iterator adapter compile-time
-// blowup issues by using a makeshift "non-determinism arrow toolkit".
-pub mod nd {
-    use std::iter;
-    use std::marker::PhantomData;
-
-    pub trait Arrow: Copy {
-        type Input;
+pub mod typed {
+    use super::{GrammarReflector, MoreThanOne, Node, ParseForest};
+    use crate::input::Input;
+
+    #[derive(Clone)]
+    enum Void {}
+
+    // HACK(eddyb) this type uses `T` but is also uninhabited.
+    type PhantomVoid<T> = (Void, std::marker::PhantomData<T>);
+
+    pub trait Shaped {
+        type Shape: Shape;
+
+        // FIXME(eddyb) this is always `[usize; Self::Shape::STATE_LEN]`.
+        // (but that doesn't work yet)
+        type State: Default + AsMut<[usize]>;
+    }
+
+    pub trait FromShapeFields<'a, 'i, G: GrammarReflector, I: Input>: Sized {
         type Output;
-        type Iter: Iterator<Item = Self::Output> + Clone;
-        fn apply(&self, x: Self::Input) -> Self::Iter;
 
-        fn map<F: Fn(Self::Output) -> R, R>(self, f: F) -> Map<Self, F> {
-            Map(self, f)
-        }
-        fn then<B: Arrow<Input = Self::Output>>(self, b: B) -> Then<Self, B> {
-            Then(self, b)
+        // FIXME(eddyb) use an array length const here instead when that works.
+        type Fields: Default + AsMut<[Option<Node<'i, G>>]>;
+
+        fn from_shape_fields(
+            forest: &'a ParseForest<'i, G, I>,
+            fields: Self::Fields,
+        ) -> Self::Output;
+
+        fn one(
+            forest: &'a ParseForest<'i, G, I>,
+            node: Node<'i, G>,
+        ) -> Result<Self::Output, MoreThanOne>
+        where
+            Self: Shaped,
+        {
+            let mut state = Self::State::default();
+            let state = state.as_mut();
+            assert_eq!(state.len(), Self::Shape::STATE_LEN);
+
+            Self::Shape::init(forest, node, state);
+
+            let mut fields = Self::Fields::default();
+            Self::Shape::read(forest, node, state, fields.as_mut());
+
+            if Self::Shape::step(forest, node, state) {
+                Err(MoreThanOne)
+            } else {
+                Ok(Self::from_shape_fields(forest, fields))
```

Comment:
This is not that different from using `all`, maybe I should just get rid of `one` APIs.